firedrake/interpolation.py

```diff
@@ -480,6 +495,125 @@ def compose_map_and_cache(map1, map2):
+
+        6 ---------13----------20---------27---------34
+        |                       |                     |
+        5          12          19         26         33
+        |                       |                     |
+        4 ---------11----------18---------25---------32
+        |                       |                     |
+        3          10          17         24         31
+        |                       |                     |
+        2 ---------9-----------16---------23---------30
+        |                       |                     |
+        1          8           15         22         29
+        | base_cell_num = 0     |  base_cell_num = 1  |
+        0 ---------7-----------14---------21---------28
+
+
+    Cell node map values for an extruded mesh are indexed by the base cell
+    number (rows) and the degree of freedom (DoF) index (columns). So
+    ``extruded_cell_node_map.values[0] = [14, 15, 16,  0,  1,  2,  7,  8,  9]``
+    are all the DoF/node numbers for the ``base_cell_num = 0``.
+    Similarly
+    ``extruded_cell_node_map.values[1] = [28, 29, 30, 21, 22, 23, 14, 15, 16]``
+    contain all 9 of the DoFs for ``base_cell_num = 1``.
+    To get the DoFs/nodes for the rest of the  cells we need to include the
+    ``extruded_cell_node_map.offset``, which tells us how far each cell's
+    DoFs/nodes are translated up from the first layer to the second, and
+    multiply these by the the given ``extrusion_height``. So in our example
+    ``extruded_cell_node_map.offset = [2, 2, 2, 2, 2, 2, 2, 2, 2]`` (we index
+    this with the DoF/node index - it's an array because each DoF/node in the
+    extruded mesh cell, in principal, can be offset upwards by a different
+    amount).
+    For ``base_cell_num = 0`` with ``extrusion_height = 1``
+    (``parent_cell_num = 1``) we add ``1*2 = 2`` to each of the DoFs/nodes in
+    ``extruded_cell_node_map.values[0]`` to get
+    ``extruded_cell_node_map.values[0] + 1 * extruded_cell_node_map.offset[0] =
+    [16, 17, 18,  2,  3,  4,  9, 10, 11]`` where ``0`` is the DoF/node index.
+
+    For each cell (vertex) of a vertex only mesh immersed in a parent
+    extruded mesh, we can can get the corresponding ``base_cell_num`` and
+    ``extrusion_height`` of the parent extruded mesh. Armed with this
+    information we use the above to work out the corresponding DoFs/nodes on
+    the parent extruded mesh.
+
+    """
+    if not isinstance(vertex_only_mesh.topology, firedrake.mesh.VertexOnlyMeshTopology):
+        raise TypeError("The input mesh must be a VertexOnlyMesh")
+    cnm = extruded_cell_node_map
+    vmx = vertex_only_mesh
+    dofs_per_target_cell = cnm.arity
+    base_cells = vmx.cell_parent_base_cell_list
+    heights = vmx.cell_parent_extrusion_height_list
+    target_cell_parent_node_list = [
+        [
+            cnm.values[base_cell, dof_idx] + height * cnm.offset[dof_idx]
+            for dof_idx in range(dofs_per_target_cell)
+        ]
```

Comment:
```suggestion
        cnm.values[base_cell, :] + height * cnm.offset[:]
```

Reply:
And then you can maybe do more broadcasting as:
```
cnm.values[base_cells, :] + heights * cnm.offset[:]
```
?

Might need to use numpy.tile or similar.

Reply:
Agreed on your suggestion, but I think going any further loses readability